src/utils.jl

```diff
@@ -294,6 +294,164 @@ function unique_ops!(ops)
+end
+
+_to_numeric(op::Destroy, b::QuantumOpticsBase.FockBasis; kwargs...) = QuantumOpticsBase.destroy(b)
+_to_numeric(op::Create, b::QuantumOpticsBase.FockBasis; kwargs...) = QuantumOpticsBase.create(b)
+function _to_numeric(op::Transition, b::QuantumOpticsBase.NLevelBasis; kwargs...)
+    i, j = _convert_levels(op; kwargs...)
+    return QuantumOpticsBase.transition(b, i, j)
+end
+
+function _convert_levels(op; level_map = nothing)
+    i, j = op.i, op.j
+    if level_map === nothing
+        if (!(i isa Number) || !(j isa Number))
+            throw(ArgumentError("Mapping from symbolic levels $(i) and $(j) to NLevelBasis requires kwarg level_map to be set"))
+        end
+        return op.i, op.j  # assume mapping between integers is just equal
+    else
+        i = level_map[op.i]
+        j = level_map[op.j]
+        return i, j
+    end
+end
+
+check_basis_match(h, b) = throw(ArgumentError("Hilbert space $h and basis $b are incompatible!"))
+check_basis_match(::FockSpace, ::QuantumOpticsBase.FockBasis) = nothing
+function check_basis_match(h::NLevelSpace, b::QuantumOpticsBase.NLevelBasis)
+    if length(h.levels) != length(b)
+        throw(ArgumentError("Hilbert space $h and basis $b have incompatible levels!"))
+    end
+end
+
+function check_basis_match(h::ProductSpace, b::QuantumOpticsBase.CompositeBasis)
+    length(h.spaces) == length(b.bases) || throw(ArgumentError("Hilbert space $h and basis $b don't have the same number of subspaces!"))
+    for (h_, b_) ∈ zip(h.spaces, b.bases)
+        check_basis_match(h_, b_)
+    end
+end
+
+# function check_basis_match
+
+# Composite bases
+function to_numeric(op::QSym, b::QuantumOpticsBase.CompositeBasis; kwargs...)
+    check_basis_match(op.hilbert, b)
+    aon = acts_on(op)
+    op_num = _to_numeric(op, b.bases[aon]; kwargs...)
+    return QuantumOpticsBase.embed(b, aon, op_num)
+end
+
+# Symbolic expressions
+function to_numeric(op::QTerm, b::QuantumOpticsBase.Basis; kwargs...)
+    f = SymbolicUtils.operation(op)
+    args = SymbolicUtils.arguments(op)
+    if f === (*)
+        if isone(args[1])
+            deleteat!(args, 1)
+        end
+        args_num = [to_numeric(arg, b; kwargs...) for arg ∈ args]
+        op_num = one(b)
+        tmp = QuantumOpticsBase.SparseOperator(b)
+        for arg ∈ args_num
+            mul!(tmp, op_num, arg)
+            op_num = tmp
+        end
+    elseif f === (+)
+        if iszero(args[1])
+            deleteat!(args, 1)
+        end
+        args_num = [to_numeric(arg, b; kwargs...) for arg ∈ args]
+        op_num = QuantumOpticsBase.SparseOperator(b)
+        for arg ∈ args_num
+            op_num.data .+= arg.data
+        end
+    elseif f === (-)
+        args_num = [to_numeric(arg, b; kwargs...) for arg ∈ args]
+        op_num = QuantumOpticsBase.SparseOperator(b)
+        for arg ∈ args_num
+            op_num.data .-= arg.data
+        end
+    else
+        args_num = [to_numeric(arg, b; kwargs...) for arg ∈ args]
```

Comment:
Same as in line 395, if it should not happen I would delete this `else`.

Reply:
Let's keep this fallback. It will handle the `+,-` cases and should work for any function that someone defines, which makes things more generic.